src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java

```diff
@@ -0,0 +1,45 @@
+package uk.gov.pay.directdebit.mandate.services.gocardless;
+
+import uk.gov.pay.directdebit.events.dao.GoCardlessEventDao;
+import uk.gov.pay.directdebit.gatewayaccounts.model.GoCardlessOrganisationId;
+import uk.gov.pay.directdebit.mandate.model.GoCardlessMandateId;
+import uk.gov.pay.directdebit.mandate.model.MandateState;
+import uk.gov.pay.directdebit.payments.model.GoCardlessEvent;
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static uk.gov.pay.directdebit.mandate.model.MandateState.ACTIVE;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.CANCELLED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.CREATED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.FAILED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.SUBMITTED;
+
+public class GoCardlessMandateStateCalculator {
+
+    private final GoCardlessEventDao goCardlessEventDao;
+
+    private static final Map<String, MandateState> GOCARDLESS_ACTION_TO_MANDATE_STATE = Map.of(
+            "created", CREATED,
+            "submitted", SUBMITTED,
+            "active", ACTIVE,
+            "cancelled", CANCELLED,
+            "failed", FAILED
+    );
+
+    public static final List<String> GOCARDLESS_ACTIONS_THAT_CHANGE_STATE = List.copyOf(GOCARDLESS_ACTION_TO_MANDATE_STATE.keySet());
+
+    @Inject
+    public GoCardlessMandateStateCalculator(GoCardlessEventDao goCardlessEventDao) {
+        this.goCardlessEventDao = goCardlessEventDao;
+    }
+
+    public Optional<MandateState> calculate(GoCardlessMandateId goCardlessMandateId, GoCardlessOrganisationId goCardlessOrganisationId) {
+        return goCardlessEventDao.findLatestApplicableEventForMandate(goCardlessMandateId, goCardlessOrganisationId, GOCARDLESS_ACTIONS_THAT_CHANGE_STATE)
```

Comment:
Made the comment on the Dao but perhaps we could use your new `GoCardlessMandateIdAndOrganisationId`?